File: src/test/java/no/nav/sosialhjelp/soknad/web/rest/ressurser/inntekt/SkattbarInntektRessursTest.java
package no.nav.sosialhjelp.soknad.web.rest.ressurser.inntekt;

import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomi;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.JsonOkonomiopplysninger;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomiOpplysningUtbetaling;
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomibekreftelse;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository;
import no.nav.sosialhjelp.soknad.business.service.TextService;
import no.nav.sosialhjelp.soknad.business.service.systemdata.SkattetatenSystemdata;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.model.exception.AuthorizationException;
import no.nav.sosialhjelp.soknad.domain.model.oidc.StaticSubjectHandlerService;
import no.nav.sosialhjelp.soknad.domain.model.oidc.SubjectHandler;
import no.nav.sosialhjelp.soknad.web.rest.ressurser.inntekt.SkattbarInntektRessurs.SkattbarInntektFrontend;
import no.nav.sosialhjelp.soknad.web.sikkerhet.Tilgangskontroll;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_SKATTEETATEN;
import static no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE;
import static no.nav.sosialhjelp.soknad.business.mappers.OkonomiMapper.setBekreftelse;
import static no.nav.sosialhjelp.soknad.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SkattbarInntektRessursTest {

    private static final String BEHANDLINGSID = "123";
    private static final String EIER = "123456789101";

    @Mock
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @Mock
    private Tilgangskontroll tilgangskontroll;

    @Mock
    private SkattetatenSystemdata skattetatenSystemdata;

    @Mock
    private TextService textService;

    @InjectMocks
    private SkattbarInntektRessurs skattbarInntektRessurs;

    @BeforeEach
    public void setUp() {
        System.setProperty("environment.name", "test");
        SubjectHandler.setSubjectHandlerService(new StaticSubjectHandlerService());
    }

    @AfterEach
    public void tearDown() {
        SubjectHandler.resetOidcSubjectHandlerService();
        System.clearProperty("environment.name");
    }

    @Test
    void getSkattbarInntektSkalReturnereTomListe() {
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER)));

        SkattbarInntektFrontend skattbarInntektFrontend = skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID);

        assertThat(skattbarInntektFrontend.inntektFraSkatteetaten).isEmpty();
    }

    @Test
    void getSkattbarInntektSkalReturnereBekreftetSkattbarInntekt() {
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithSkattbarInntekt(true));

        SkattbarInntektFrontend skattbarInntektFrontend = skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID);

        assertThat(skattbarInntektFrontend.inntektFraSkatteetaten).hasSize(1);
    }

    @Test
    void getSkattbarInntektSkalReturnereHarIkkeSkattbarInntekt() {
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(
                createJsonInternalSoknadWithSkattbarInntekt(false));

        SkattbarInntektFrontend skattbarInntektFrontend = skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID);

        assertThat(skattbarInntektFrontend.inntektFraSkatteetaten).isEmpty();
    }

    @Test
    void skattbarInntekt_skalGiSamtykke() {
        SoknadUnderArbeid soknad = createJsonInternalSoknadWithSkattbarInntekt(false);
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknad);
        when(textService.getJsonOkonomiTittel(anyString())).thenReturn("tittel");
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(any());

        skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, true, "token");

        // Sjekker kaller til bostotteSystemdata
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(skattetatenSystemdata).updateSystemdataIn(argument.capture());
        JsonOkonomi okonomi = argument.getValue().getJsonInternalSoknad().getSoknad().getData().getOkonomi();
        JsonOkonomibekreftelse fangetBekreftelse = okonomi.getOpplysninger().getBekreftelse().get(0);
        assertThat(fangetBekreftelse.getType()).isEqualTo(UTBETALING_SKATTEETATEN_SAMTYKKE);
        assertThat(fangetBekreftelse.getVerdi()).isTrue();

        // Sjekker lagring av soknaden
        SoknadUnderArbeid spartSoknad = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        assertThat(spartSoknad.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger().getBekreftelse()).hasSize(1);
        JsonOkonomibekreftelse spartBekreftelse = soknad.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger().getBekreftelse().get(0);
        assertThat(spartBekreftelse.getType()).isEqualTo(UTBETALING_SKATTEETATEN_SAMTYKKE);
        assertThat(spartBekreftelse.getVerdi()).isTrue();
    }

    @Test
    void skattbarInntekt_skalTaBortSamtykke() {
        SoknadUnderArbeid soknad = createJsonInternalSoknadWithSkattbarInntekt(false);
        JsonOkonomiopplysninger opplysninger = soknad.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger();
        setBekreftelse(opplysninger, UTBETALING_SKATTEETATEN_SAMTYKKE, true, "");
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknad);
        when(textService.getJsonOkonomiTittel(anyString())).thenReturn("tittel");
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(any());

        skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, false, "token");

        // Sjekker kaller til skattbarInntektSystemdata
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(skattetatenSystemdata).updateSystemdataIn(argument.capture());
        JsonOkonomi okonomi = argument.getValue().getJsonInternalSoknad().getSoknad().getData().getOkonomi();
        JsonOkonomibekreftelse fangetBekreftelse = okonomi.getOpplysninger().getBekreftelse().get(0);
        assertThat(fangetBekreftelse.getType()).isEqualTo(UTBETALING_SKATTEETATEN_SAMTYKKE);
        assertThat(fangetBekreftelse.getVerdi()).isFalse();

        // Sjekker lagring av soknaden
        SoknadUnderArbeid spartSoknad = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        JsonOkonomiopplysninger sparteOpplysninger = spartSoknad.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger();
        assertThat(sparteOpplysninger.getBekreftelse()).hasSize(1);
        JsonOkonomibekreftelse spartBekreftelse = sparteOpplysninger.getBekreftelse().get(0);
        assertThat(spartBekreftelse.getType()).isEqualTo(UTBETALING_SKATTEETATEN_SAMTYKKE);
        assertThat(spartBekreftelse.getVerdi()).isFalse();
    }

    @Test
    void skattbarInntekt_skalIkkeForandreSamtykke() {
        SoknadUnderArbeid soknad = createJsonInternalSoknadWithSkattbarInntekt(false);
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknad);
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(any());

        skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, false, "token");

        // Sjekker kaller til skattbarInntektSystemdata
        verify(skattetatenSystemdata, times(0)).updateSystemdataIn(any());

        // Sjekker lagring av soknaden
        verify(soknadUnderArbeidRepository, times(0)).oppdaterSoknadsdata(any(), anyString());

        // Sjekker soknaden
        assertThat(soknad.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger().getBekreftelse()).isEmpty();
    }

    @Test
    void getSkattbarInntektkalKasteAuthorizationExceptionVedManglendeTilgang() {
        doThrow(new AuthorizationException("Not for you my friend")).when(tilgangskontroll).verifiserAtBrukerHarTilgang();

        assertThatExceptionOfType(AuthorizationException.class)
                .isThrownBy(() -> skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID));

        verifyNoInteractions(soknadUnderArbeidRepository);
    }

    @Test
    void putSamtykkeSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        doThrow(new AuthorizationException("Not for you my friend")).when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(BEHANDLINGSID);

        assertThatExceptionOfType(AuthorizationException.class)
                .isThrownBy(() -> skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, true, "token"));

        verifyNoInteractions(soknadUnderArbeidRepository);
    }

    private SoknadUnderArbeid catchSoknadUnderArbeidSentToOppdaterSoknadsdata() {
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(soknadUnderArbeidRepository).oppdaterSoknadsdata(argument.capture(), anyString());
        return argument.getValue();
    }

    private SoknadUnderArbeid createJsonInternalSoknadWithSkattbarInntekt(Boolean harSkattbarInntekt) {
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        if (harSkattbarInntekt) {
            JsonOkonomiOpplysningUtbetaling utbetaling = new JsonOkonomiOpplysningUtbetaling()
                    .withType(UTBETALING_SKATTEETATEN)
                    .withKilde(JsonKilde.SYSTEM)
                    .withTittel("Utbetalingen!")
                    .withBelop(123456);
            soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getOkonomi().getOpplysninger().getUtbetaling().add(utbetaling);
        }
        return soknadUnderArbeid;
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/inntekt/skattbarinntekt/SkattbarInntektRessursTest.kt
package no.nav.sosialhjelp.soknad.inntekt.skattbarinntekt

import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.runs
import io.mockk.slot
import io.mockk.unmockkObject
import io.mockk.verify
import no.nav.sbl.soknadsosialhjelp.json.SoknadJsonTyper
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sbl.soknadsosialhjelp.soknad.okonomi.opplysning.JsonOkonomiOpplysningUtbetaling
import no.nav.sosialhjelp.soknad.app.MiljoUtils
import no.nav.sosialhjelp.soknad.app.exceptions.AuthorizationException
import no.nav.sosialhjelp.soknad.app.mapper.OkonomiMapper
import no.nav.sosialhjelp.soknad.app.subjecthandler.StaticSubjectHandlerImpl
import no.nav.sosialhjelp.soknad.app.subjecthandler.SubjectHandlerUtils
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.SoknadService.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.tekster.TextService
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class SkattbarInntektRessursTest {

    private val soknadUnderArbeidRepository: SoknadUnderArbeidRepository = mockk()
    private val tilgangskontroll: Tilgangskontroll = mockk()
    private val skatteetatenSystemdata: SkatteetatenSystemdata = mockk()
    private val textService: TextService = mockk()
    private val skattbarInntektRessurs = SkattbarInntektRessurs(
        tilgangskontroll,
        soknadUnderArbeidRepository,
        skatteetatenSystemdata,
        textService
    )

    @BeforeEach
    fun setUp() {
        mockkObject(MiljoUtils)
        every { MiljoUtils.isNonProduction() } returns true
        SubjectHandlerUtils.setNewSubjectHandlerImpl(StaticSubjectHandlerImpl())
    }

    @AfterEach
    fun tearDown() {
        SubjectHandlerUtils.resetSubjectHandlerImpl()
        unmockkObject(MiljoUtils)
    }

    @Test
    fun skattbarInntektSkalReturnereTomListe() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns createSoknadUnderArbeid()

        val skattbarInntektFrontend = skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID)
        assertThat(skattbarInntektFrontend.inntektFraSkatteetaten).isEmpty()
    }

    @Test
    fun skattbarInntektSkalReturnereBekreftetSkattbarInntekt() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every {
            soknadUnderArbeidRepository.hentSoknad(any<String>(), any())
        } returns createJsonInternalSoknadWithSkattbarInntekt(true)

        val skattbarInntektFrontend = skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID)
        assertThat(skattbarInntektFrontend.inntektFraSkatteetaten).hasSize(1)
    }

    @Test
    fun skattbarInntektSkalReturnereHarIkkeSkattbarInntekt() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } just runs
        every {
            soknadUnderArbeidRepository.hentSoknad(any<String>(), any())
        } returns createJsonInternalSoknadWithSkattbarInntekt(false)

        val skattbarInntektFrontend = skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID)
        assertThat(skattbarInntektFrontend.inntektFraSkatteetaten).isEmpty()
    }

    @Test
    fun skattbarInntekt_skalGiSamtykke() {
        val soknad = createJsonInternalSoknadWithSkattbarInntekt(false)
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknad
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs

        val soknadUnderArbeidSlot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(soknadUnderArbeidSlot), any()) } just runs

        val systemdataSlot = slot<SoknadUnderArbeid>()
        every { skatteetatenSystemdata.updateSystemdataIn(capture(systemdataSlot)) } just runs

        skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, true, "token")

        // Sjekker kaller til skatteetatenSystemdata
        verify { skatteetatenSystemdata.updateSystemdataIn(systemdataSlot.captured) }

        val okonomi = systemdataSlot.captured.jsonInternalSoknad!!.soknad.data.okonomi
        val fangetBekreftelse = okonomi.opplysninger.bekreftelse[0]
        assertThat(fangetBekreftelse.type).isEqualTo(SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE)
        assertThat(fangetBekreftelse.verdi).isTrue

        // Sjekker lagring av soknaden
        val spartSoknad = soknadUnderArbeidSlot.captured
        assertThat(spartSoknad.jsonInternalSoknad!!.soknad.data.okonomi.opplysninger.bekreftelse).hasSize(1)
        val spartBekreftelse = soknad.jsonInternalSoknad!!.soknad.data.okonomi.opplysninger.bekreftelse[0]
        assertThat(spartBekreftelse.type).isEqualTo(SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE)
        assertThat(spartBekreftelse.verdi).isTrue
    }

    @Test
    fun skattbarInntekt_skalTaBortSamtykke() {
        val soknad = createJsonInternalSoknadWithSkattbarInntekt(false)
        val opplysninger = soknad.jsonInternalSoknad!!.soknad.data.okonomi.opplysninger
        OkonomiMapper.setBekreftelse(opplysninger, SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE, true, "")
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknad
        every { textService.getJsonOkonomiTittel(any()) } returns "tittel"
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs

        val soknadUnderArbeidSlot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(soknadUnderArbeidSlot), any()) } just runs

        val systemdataSlot = slot<SoknadUnderArbeid>()
        every { skatteetatenSystemdata.updateSystemdataIn(capture(systemdataSlot)) } just runs

        skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, false, "token")

        // Sjekker kaller til skattbarInntektSystemdata
        verify { skatteetatenSystemdata.updateSystemdataIn(systemdataSlot.captured) }

        val okonomi = systemdataSlot.captured.jsonInternalSoknad!!.soknad.data.okonomi
        val fangetBekreftelse = okonomi.opplysninger.bekreftelse[0]
        assertThat(fangetBekreftelse.type).isEqualTo(SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE)
        assertThat(fangetBekreftelse.verdi).isFalse

        // Sjekker lagring av soknaden
        val spartSoknad = soknadUnderArbeidSlot.captured
        val sparteOpplysninger = spartSoknad.jsonInternalSoknad!!.soknad.data.okonomi.opplysninger
        assertThat(sparteOpplysninger.bekreftelse).hasSize(1)
        val spartBekreftelse = sparteOpplysninger.bekreftelse[0]
        assertThat(spartBekreftelse.type).isEqualTo(SoknadJsonTyper.UTBETALING_SKATTEETATEN_SAMTYKKE)
        assertThat(spartBekreftelse.verdi).isFalse
    }

    @Test
    fun skattbarInntektkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerHarTilgang() } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { skattbarInntektRessurs.hentSkattbareInntekter(BEHANDLINGSID) }

        verify(exactly = 0) { soknadUnderArbeidRepository.oppdaterSoknadsdata(any(), any()) }
    }

    @Test
    fun putSamtykkeSkalKasteAuthorizationExceptionVedManglendeTilgang() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(BEHANDLINGSID) } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { skattbarInntektRessurs.updateSamtykke(BEHANDLINGSID, true, "token") }

        verify(exactly = 0) { soknadUnderArbeidRepository.oppdaterSoknadsdata(any(), any()) }
    }

    private fun createJsonInternalSoknadWithSkattbarInntekt(harSkattbarInntekt: Boolean): SoknadUnderArbeid {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        if (harSkattbarInntekt) {
            val utbetaling = JsonOkonomiOpplysningUtbetaling()
                .withType(SoknadJsonTyper.UTBETALING_SKATTEETATEN)
                .withKilde(JsonKilde.SYSTEM)
                .withTittel("Utbetalingen!")
                .withBelop(123456)
            soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.okonomi.opplysninger.utbetaling.add(utbetaling)
        }
        return soknadUnderArbeid
    }

    companion object {
        private const val BEHANDLINGSID = "123"
        private const val EIER = "123456789101"

        private fun createSoknadUnderArbeid(): SoknadUnderArbeid {
            return SoknadUnderArbeid(
                versjon = 1L,
                behandlingsId = BEHANDLINGSID,
                tilknyttetBehandlingsId = null,
                eier = EIER,
                jsonInternalSoknad = createEmptyJsonInternalSoknad(EIER),
                status = SoknadUnderArbeidStatus.UNDER_ARBEID,
                opprettetDato = LocalDateTime.now(),
                sistEndretDato = LocalDateTime.now()
            )
        }
    }
}
